cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/ValidationUtils.java

```diff
@@ -0,0 +1,88 @@
+import io.cdap.cdap.etl.batch.BatchPipelineSpecGenerator;
+import io.cdap.cdap.etl.common.DefaultPipelineConfigurer;
+import io.cdap.cdap.etl.common.DefaultStageConfigurer;
+import io.cdap.cdap.etl.proto.v2.ETLBatchConfig;
+import io.cdap.cdap.etl.proto.v2.ETLPlugin;
+import io.cdap.cdap.etl.proto.v2.ETLStage;
+import io.cdap.cdap.etl.proto.v2.spec.StageSpec;
+import io.cdap.cdap.etl.proto.v2.validation.StageSchema;
+import io.cdap.cdap.etl.proto.v2.validation.StageValidationRequest;
+import io.cdap.cdap.etl.proto.v2.validation.StageValidationResponse;
+import io.cdap.cdap.etl.spec.PipelineSpecGenerator;
+import io.cdap.cdap.etl.validation.ValidatingConfigurer;
+import io.cdap.cdap.internal.io.SchemaTypeAdapter;
+
+import java.util.Collections;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+/**
+ * Utility functions for common pipeline validation code
+ */
+public class ValidationUtils {
+
+  private static final Gson GSON = new GsonBuilder()
+    .registerTypeAdapter(Schema.class, new SchemaTypeAdapter())
+    .serializeNulls()
+    .create();
+
+  public static String validate(StageValidationRequest validationRequest,
+                                PluginConfigurer pluginConfigurer,
+                                Function<Map<String, String>, Map<String, String>> macroFn) {
+
+    ETLStage stageConfig = validationRequest.getStage();
+    ValidatingConfigurer validatingConfigurer = new ValidatingConfigurer(pluginConfigurer);
+    // Batch or Streaming doesn't matter for a single stage.
+    PipelineSpecGenerator<ETLBatchConfig, BatchPipelineSpec> pipelineSpecGenerator =
+      new BatchPipelineSpecGenerator(validatingConfigurer, Collections.emptySet(), Collections.emptySet(),
+                                     Engine.SPARK);
+
+    DefaultStageConfigurer stageConfigurer = new DefaultStageConfigurer(stageConfig.getName());
+    for (StageSchema stageSchema : validationRequest.getInputSchemas()) {
+      stageConfigurer.addInputSchema(stageSchema.getStage(), stageSchema.getSchema());
+      stageConfigurer.addInputStage(stageSchema.getStage());
+    }
+    DefaultPipelineConfigurer pipelineConfigurer =
+      new DefaultPipelineConfigurer(validatingConfigurer, stageConfig.getName(), Engine.SPARK, stageConfigurer);
+
+    // evaluate macros
+    Map<String, String> evaluatedProperties = macroFn.apply(stageConfig.getPlugin().getProperties());
+
+    ETLPlugin originalConfig = stageConfig.getPlugin();
+    ETLPlugin evaluatedConfig = new ETLPlugin(originalConfig.getName(), originalConfig.getType(),
+                                              evaluatedProperties, originalConfig.getArtifactConfig());
+    try {
+      StageSpec spec = pipelineSpecGenerator.configureStage(stageConfig.getName(), evaluatedConfig,
+                                                            pipelineConfigurer).build();
+      return GSON.toJson(new StageValidationResponse(spec));
```

Comment:
Generally not a good ideal using a Json string as the method contract. Json serialization should only happen right before writing out the data structure. Why need to do the serialization here?

Reply:
Changed method to return StageValidationResponse